views/course_form/views.py

```diff
@@ -23,8 +61,108 @@ def post(self, request, code: str | None = None, semester: str | None = None):
+        elif not re.match(r'^[A-Za-z0-9]+$', code):
+            errors.append("Course code must be a valid value")  # Same error message for consistency
+
+            # validate name
+        if not name:
+            errors.append('Course name must be a valid value')
+        elif not re.match(r'^[A-Za-z0-9 ]+$', name):  # Allow alphanumeric and spaces
+            errors.append('Course name must be a valid value')
+
+        # Validate semester
+        if not CourseController.CourseController.validate_semester(semester):
+            return HttpResponse(
+                repr(["Course semester must be a valid semester"]),
+                content_type="text/plain",
+                status=200
+            )
+
+        if not CourseController.CourseController.validate_semester(semester):
+            errors.append('Course semester must be a valid semester')
+
+            # Validate TA list
+        if ta_list:
+            ta_usernames = [ta.strip() for ta in ta_list.split(",")]
+            non_existent_tas = [
+                ta for ta in ta_usernames if not User.objects.filter(username=ta).exists()
+            ]
+            if non_existent_tas:
+                errors.append('One or more of the selected TAs is not an existing user')
+
+        # TA Role validation
+        if ta_list:
+            ta_usernames = [ta.strip() for ta in ta_list.split(",")]
+            invalid_tas = [
+                ta for ta in ta_usernames
+                if not User.objects.filter(username=ta, role="TA").exists()
+            ]
+            if invalid_tas:
+                errors.append('One or more of the selected TAs is not a TA')
+
+        # If validation fails, return errors in the response
+        if errors:
+            print(errors)
+            return HttpResponse(repr(errors), content_type="text/plain", status=200)
+
+            # Check if the course exists
+        try:
+            course = Course.objects.get(course_code=code)
+            # Check if the semester is being changed
+            if course.semester.semester_name != semester:
+                errors.append("Cannot change course semester")
+        except Course.DoesNotExist:
+            course = None  # For new courses, allow creation
+
+        semester_instance = Semester.objects.get(semester_name=semester)
+        # Create course data and call save_course
+        try:
+            course_data = CourseFormData(
+                course_code=code,
+                course_name=name,
+                semester=semester,
+                ta_username_list=ta_list,
+            )
+            course_id = CourseController.CourseController.save_course(course_data)
+        except ValueError as e:
+            print(f"Error saving course: {e}")
+            return HttpResponse(str(e), status=400)
+
+        # Assign TAs to the course
+        for ta in ta_list:
+            TACourseAssignment.objects.create(ta=ta, course_id=course_id, grader_status=False)
```

Comment:
Shouldn't do DB access.
It looks like the course controller is missing the functionality for TA course assignments. Let if fail for now and create another story to test and implement.